### DISCUSSION TITLE
[@types/react] Inconsistent assignability of ComponentClass and FunctionComponent

### DISCUSSION THREAD

**Discussion by joshuakb2** [DefinitelyTyped/DefinitelyTyped#64154] [Issues with a @types package]
I found a behavior that surprised me regarding the `ComponentType` type, which describes a component with a particular set of props and state that may either be a class component or a functional component. It seems that `ComponentType<Props1> extends ComponentType<Props2>` is only true if `Props1` and `Props2` are equivalent. This is because function components and class components have inverse assignability behavior regarding props.

Here's a sandbox that demonstrates what I mean: https://codesandbox.io/s/assignability-of-react-class-components-vs-function-components-1bgvb6?file=/src/index.tsx

To sum it up here, `FunctionComponent<Props1> extends FunctionComponent<Props2>` is true if `Props2 extends Props1` is also true. This makes sense because components take props as input. However, `ComponentClass<Props1> extends ComponentClass<Props2>` is only true if `Props1 extends Props2`. This seems wrong because functional components and class components are meant to be interchangeable from the user's perspective.

Is this behavior intentional?

[a comment by an automated account is not shown]
